### PR TITLE
Add short-circuiting for painless def equality

### PR DESCRIPTION
**Pull request by thecoop**
Add short-circuiting checks to `Def.eq` implementation.
Remove some obsoleted code for array handles and string concat.

This shows a 8% increase in throughput in the index-append-with-ingest-baseline-pipeline rally test

### CODE REVIEW

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/DefMath.java [hunk 1]
```diff
@@ -526,59 +526,63 @@ private static boolean eq(boolean a, boolean b) {
...
-            if (left instanceof Double) {
-                if (right instanceof Number) {
-                    return (double) left == ((Number) right).doubleValue();
-                } else if (right instanceof Character) {
-                    return (double) left == (char) right;
+        if (left == right) {
+            return true;
+        } else if (left != null && right != null) {
+            if (left.getClass() == right.getClass()) {
+                return left.equals(right);
```
Comment (thecoop) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/DefMath.java:533]: These lines are the new ones to short-circuit when they're the same instance or same type

**Comment by thecoop** [comment]
@elasticsearchmachine rerun elasticsearch-ci/packaging-tests-windows-sample

**Comment by stu-elastic** [comment]
> Remove some obsoleted code for array handles and string concat.

Please pull these into separate PRs, it's nice to keep PRs focused on one task.

**Comment by thecoop** [comment]
Created #92102 just for the short-circuit change

**Comment by thecoop** [comment]
Moved refactorings into #92103

**Pull request closed without merging**